hazelcast-jet-core/src/main/java/com/hazelcast/jet/core/CloseableProcessorSupplier.java

```diff
@@ -57,13 +65,27 @@ public CloseableProcessorSupplier(DistributedSupplier<E> simpleSupplier) {
     }
 
     /**
-     * @param supplier Supplier to create processor instances.
-     *                 Parameter is the number of processors that should be in the collection.
+     * Constructs an instance which will wrap all the processors the provided
+     * {@code supplier} returns.
+     *
+     * @param supplier supplier of processors. The {@code int} parameter passed to it is the
+     *                 number of processors that should be in the returned collection.
      */
     public CloseableProcessorSupplier(DistributedIntFunction<Collection<E>>  supplier) {
         this.supplier = supplier;
     }
 
+    /**
+     * Initializes this object with the given supplier. May only be invoked if
+     * the supplier property hasn't been enitialized yet.
+     */
+    public void setSupplier(DistributedIntFunction<Collection<E>> newSupplier) {
```

Comment:
I don't quite like this.. why not have the supplier take as constructor `Function<JetInstance, E>` instead? or `Function<Context, E>` ?

Reply:
It will need to be `Function<JetInstance, Supplier<E>>` because I need to create the shared context based on JetInstance and share it for all processor instances. I wanted to avoid supplier of supplier.

Reply:
I think it's fine, I find this code hard to follow and it requires that we extend the class rather than just using it.

Reply:
yes.. We might refactor everything to use setSupplier. Super-constructor call is bad in that you can't use any local state in it.

Reply:
refactored it to only have `setSupplier`, it's simpler now